qiskit/pulse/builder.py

```diff
@@ -0,0 +1,2000 @@
+
+
+@contextmanager
+def phase_offset(phase: float,
+                 *channels: chans.PulseChannel
+                 ) -> ContextManager[None]:
+    """Shift the phase of input channels on entry into context and undo on exit.
+
+    Examples:
+
+    .. jupyter-execute::
+
+        import math
+
+        from qiskit import pulse
+
+        d0 = pulse.DriveChannel(0)
+
+        with pulse.build(backend) as pulse_prog:
+            with pulse.phase_offset(math.pi, d0):
+                pulse.play(pulse.Constant(10, 1.0), d0)
+
+        assert len(pulse_prog.instructions) == 3
+
+    Args:
+        phase: Amount of phase offset in radians.
+        channels: Channels to offset phase of.
+
+    Yields:
+        None
+    """
+    for channel in channels:
+        shift_phase(phase, channel)
+    try:
+        yield
+    finally:
+        for channel in channels:
+            shift_phase(-phase, channel)
+
+
+@contextmanager
+def frequency_offset(frequency: float,
+                     channels: chans.PulseChannel,
+                     compensate_phase: bool = False
+                     ) -> ContextManager[None]:
+    """Shift the frequency of inputs channels on entry into context and undo on exit.
+
+    Examples:
+
+    .. code-block:: python
+        :emphasize-lines: 7, 16
+
+        from qiskit import pulse
+
+        d0 = pulse.DriveChannel(0)
+
+        with pulse.build(backend) as pulse_prog:
+            # shift frequency by 1GHz
+            with pulse.frequency_offset(1e9, d0):
+                pulse.play(pulse.Constant(10, 1.0), d0)
+
+        assert len(pulse_prog.instructions) == 3
+
+        with pulse.build(backend) as pulse_prog:
+            # Shift frequency by 1GHz.
+            # Undo accumulated phase in the shifted freqeuncy frame
+            # when exiting the context.
+            with pulse.frequency_offset(1e9, d0, compensate_phase=True):
+                pulse.play(pulse.Constant(10, 1.0), d0)
+
+        assert len(pulse_prog.instructions) == 4
+
+    Args:
+        frequency: Amount of frequency offset in Hz.
+        channels: Channels to offset phase of.
+        compensate_phase: Compensate for accumulated phase in accumulated with
+            respect to the channels frame at its initial frequency.
+
+    Yields:
+        None
+    """
+    builder = _active_builder()
+    t0 = builder.block.duration
+
+    for channel in channels:
+        shift_frequency(frequency, channel)
+    try:
+        yield
+    finally:
+        if compensate_phase:
+            duration = builder.block.duration - t0
+            dt = active_backend().configuration().dt
+            accumulated_phase = duration * dt * frequency % (2*np.pi)
+            for channel in channels:
+                shift_phase(channel, -accumulated_phase)
+
+        for channel in channels:
+            shift_frequency(channel, -frequency)
+
+
+# Channels
+def drive_channel(qubit: int) -> chans.DriveChannel:
```

Comment:
@lcapelluto. With regards to renaming, I had considered `d` but this is too short or `dch`, but I thought this was ugly.  I am also not a fan of the shorthand acronyms such as `u` channel.